### ISSUE TITLE
Analyze the crabnet hyperparameter benchmark dataset

### ISSUE THREAD

**Issue by sgbaird** [sparks-baird/matsci-opt-benchmarks#50]
I'd like you to download the CrabNet Hyperparameter dataset from Zenodo and create a Jupyter notebook to run some analysis on it. Specifically, you'll use `sobol_regression.csv`. I'd like you to start by training a scikit-learn RFR model with and without the "rank" variables (i.e., the captured noise). If that works, also train the following:

1. Histogram Gradient Boosting (see https://scikit-learn.org/stable/auto_examples/ensemble/plot_forest_hist_grad_boosting_comparison.html#sphx-glr-auto-examples-ensemble-plot-forest-hist-grad-boosting-comparison-py)
2. SVR
3. Ridge regression
4. GPR with ARD

Here's some example code from Runze Zhang towards this end. Ignore code related to `3D printed data (Keith's data)` (a separate dataset from this repo).

```python
# -*- coding: utf-8 -*-
"""GPR_code_for_Jae_check.ipynb

Automatically generated by Colab.

Original file is located at
    https://colab.research.google.com/drive/1eGUgde7ewO_hPAM4h4raU4SZEXk-8W5g

### Data loading

#### 3D printed data (Keith's data)
"""

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR
from sklearn.model_selection import cross_val_score
from sklearn.gaussian_process.kernels import Matern, WhiteKernel, ConstantKernel as C
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler
from scipy.stats import norm
from scipy.spatial.distance import cdist
from sklearn.model_selection import cross_val_predict, KFold
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

# ignore warnings
import warnings
warnings.filterwarnings("ignore")

# load the data
printed_df = pd.read_csv('CampaignData.csv')

feature_columns = {
    "x1": "perimeter_ratio",
    "x2": "c4_base",
    "x3": "c4_top",
    "x4": "c8_base",
    "x5": "c8_top",
    "x6": "twist_linear",
    "x7": "twist_amplitude",
    "x8": "twist_cycles",
    "RecordedHeight": "height",
    "WallThickness": "thickness",
    "RecordedMass": "mass"
}

# Target column
target_column = "CriticalEfficiency"

# cleaning column
clean_column = 'Valid'

printed_df = printed_df[list(feature_columns.keys()) + [target_column] + [clean_column]].copy()

printed_df = printed_df.iloc[2:]

printed_df.head()

printed_df["Valid"] = pd.to_numeric(printed_df["Valid"], errors='coerce')

cleaned_df = printed_df[printed_df['Valid'] != 0]

cleaned_df.head()

# convert each column to numeric
for col in cleaned_df.columns:
    if col != clean_column:
        cleaned_df[col] = pd.to_numeric(cleaned_df[col], errors='coerce')

"""#### Sterling's data"""

st_df = pd.read_csv('Master Dataset.csv')
st_df.head()

# 23 features used in the dataset

feature_cols = ['N', 'alpha', 'd_model',
       'dim_feedforward', 'dropout', 'emb_scaler', 'eps', 'epochs_step',
       'fudge', 'heads', 'k', 'lr', 'pe_resolution', 'ple_resolution',
       'pos_scaler', 'weight_decay', 'batch_size', 'out_hidden4', 'betas1',
       'betas2', 'bias', 'criterion', 'elem_prop']

target_column = 'mae'

cleaned_df = st_df[feature_cols + [target_column]].copy()

# bias_ohe = pd.get_dummies(sobol_reg["bias"], prefix="bias")
criterion_ohe = pd.get_dummies(cleaned_df["criterion"], prefix="criterion")
elemprop_ohe = pd.get_dummies(cleaned_df["elem_prop"], prefix="elem_prop")
cleaned_df["bias"] = cleaned_df["bias"].astype(int)

sobol_reg = pd.concat([cleaned_df, criterion_ohe, elemprop_ohe], axis=1)

sobol_reg.drop(columns=["criterion", "elem_prop"], inplace=True)

sobol_reg

"""### Train a ML model"""

# train a random forest with 5cv fold and hyperparameter tuning
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict
from sklearn.model_selection import KFold
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.metrics import median_absolute_error
from sklearn.metrics import mean_squared_log_error
from sklearn.metrics import r2_score

"""Keith's data"""

X = cleaned_df[list(feature_columns.keys())].values
y = cleaned_df[target_column].values

kb_0005 = cleaned_df.sample(frac=0.005, random_state=42) # 66 samples

X = kb_0005[list(feature_columns.keys())].values # I know that I shouldn't use the same variable name, but I am lazy
y = kb_0005[target_column].values

"""Sterling's data"""

X = sobol_reg.iloc[:,:-1].values
y = sobol_reg['mae'].values

st_00005 = sobol_reg.sample(frac=0.0005, random_state=42) # random sampling - 87 samples
X = st_00005.iloc[:,:-1].values
y = st_00005['mae'].values

st_000005 = sobol_reg.sample(frac=0.00005, random_state=42) # random sampling - 9 samples
X = st_000005.iloc[:,:-1].values
y = st_000005['mae'].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the Random Forest Regressor
rf = RandomForestRegressor(random_state=42)
# Define the parameter grid for hyperparameter tuning
param_grid = {
    'n_estimators': [10, 50, 100],
    'max_depth': [None, 5, 10],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2],
    'max_features': ['auto', 'sqrt']
}

# Perform Grid Search with Cross-Validation
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid,
                           scoring='neg_mean_squared_error', cv=5, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)

# Get the best parameters
best_params = grid_search.best_params_
print("Best Parameters:", best_params)

# Train the Random Forest Regressor with the best parameters
best_rf = RandomForestRegressor(**best_params, random_state=42)
best_rf.fit(X_train, y_train)

# Make predictions on the test set
y_pred = best_rf.predict(X_test)

# Evaluate the model
r2 = r2_score(y_test, y_pred)

# plot parity plot
plt.figure(figsize=(5, 5))
plt.scatter(y_test, y_pred, alpha=0.7)
plt.plot([min(y_test), max(y_test)], [min(y_test), max(y_test)], 'r--', label= 'R2 = {:.2f}'.format(r2))
plt.xlabel("True Efficiency")
plt.ylabel("Predicted Efficiency")
plt.title("Parity Plot - Random Forest")
# add r2 into legend
# plt.text(0.5, 0.1, f'R² = {r2:.2f}', fontsize=12, ha='center', va='center', transform=plt.gca().transAxes)
plt.grid(True)
plt.tight_layout()
plt.legend()
plt.show()

from sklearn.svm import SVR
from sklearn.pipeline import Pipeline

# Define pipeline
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('svr', SVR(kernel='rbf'))
])

# Define parameter grid
param_grid = {
    'svr__C': [0.1, 1, 10],
    'svr__gamma': ['scale', 'auto'],
    'svr__epsilon': [0.1, 0.2, 0.5]
}

# Grid Search with pipeline
grid_search = GridSearchCV(estimator=pipeline, param_grid=param_grid,
                           scoring='neg_mean_squared_error', cv=5, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)

# Get best model
best_model = grid_search.best_estimator_

# Predict on test set
y_pred = best_model.predict(X_test)

# Evaluate
r2 = r2_score(y_test, y_pred)
print("Best Parameters:", grid_search.best_params_)

# Plot parity
plt.figure(figsize=(5, 5))
plt.scatter(y_test, y_pred, alpha=0.7)
plt.plot([min(y_test), max(y_test)], [min(y_test), max(y_test)], 'r--', label=f'R² = {r2:.2f}')
plt.xlabel("True Efficiency")
plt.ylabel("Predicted Efficiency")
plt.title("Parity Plot - SVR")
plt.grid(True)
plt.tight_layout()
plt.legend()
plt.show()

from sklearn.pipeline import Pipeline
from sklearn.linear_model import Ridge

# Define pipeline
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('ridge', Ridge())
])

# Define parameter grid (note the prefix "ridge__")
param_grid = {
    'ridge__alpha': [0.1, 1, 10],
    'ridge__solver': ['auto', 'svd', 'cholesky', 'lsqr', 'sag']
}

# Perform Grid Search with CV on raw X_train
grid_search = GridSearchCV(estimator=pipeline, param_grid=param_grid,
                           scoring='neg_mean_squared_error', cv=5, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)

# Get the best model (includes scaling!)
best_model = grid_search.best_estimator_

# Predict on raw X_test (scaling is handled inside the pipeline)
y_pred = best_model.predict(X_test)

# Evaluate the model
r2 = r2_score(y_test, y_pred)
print("Best Parameters:", grid_search.best_params_)

# Plot parity plot
plt.figure(figsize=(5, 5))
plt.scatter(y_test, y_pred, alpha=0.7)
plt.plot([min(y_test), max(y_test)], [min(y_test), max(y_test)], 'r--', label=f'R2 = {r2:.2f}')
plt.xlabel("True Efficiency")
plt.ylabel("Predicted Efficiency")
plt.title("Parity Plot - Ridge Regression")
plt.grid(True)
plt.tight_layout()
plt.legend()
plt.show()

from sklearn.pipeline import Pipeline
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF, ConstantKernel as C, WhiteKernel, Matern

# Define pipeline
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('gpr', GaussianProcessRegressor(normalize_y=True))
])

# Define parameter grid
param_grid = {
    'gpr__alpha': [1e-10, 1e-5, 1e-3, 1e-1, 1e0],
    'gpr__kernel': [
        C(1.0, (1e-3, 1e3)) * Matern(nu=2.5) + WhiteKernel(),
        C(1.0, (1e-3, 1e3)) * RBF() + WhiteKernel()
    ]
}

# GridSearchCV with pipeline
grid_search = GridSearchCV(estimator=pipeline, param_grid=param_grid,
                           scoring='neg_mean_squared_error', cv=5, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)

# Best estimator includes scaler + GPR
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)

# Evaluate
r2 = r2_score(y_test, y_pred)
print(f"Best Parameters: {grid_search.best_params_}")
print(f"R² Score: {r2:.3f}")

# Plot
plt.figure(figsize=(5, 5))
plt.scatter(y_test, y_pred, alpha=0.7)
plt.plot([min(y_test), max(y_test)], [min(y_test), max(y_test)], 'r--', label=f'R2 = {r2:.2f}')
plt.xlabel("True Efficiency")
plt.ylabel("Predicted Efficiency")
plt.title("Parity Plot - GPR")
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()

"""Do nested CV when only sampling 0.005% from Sterling's data"""

from sklearn.pipeline import Pipeline
from sklearn.model_selection import LeaveOneOut, GridSearchCV
from tqdm import tqdm

# Define hyperparameter grid
param_grid = {
    'gpr__alpha': [1e-10, 1e-5, 1e-3, 1e-1, 1e0],
    'gpr__kernel': [
        C(1.0, (1e-3, 1e3)) * Matern(nu=2.5) + WhiteKernel(),
        C(1.0, (1e-3, 1e3)) * RBF() + WhiteKernel()
    ]
}

# Outer CV: Leave-One-Out
loo = LeaveOneOut()
y_true_all = []
y_pred_all = []

for train_idx, test_idx in tqdm(loo.split(X_train), total=X_train.shape[0], desc="Nested LOOCV"):
    X_tr, X_te = X_train[train_idx], X_train[test_idx]
    y_tr, y_te = y_train[train_idx], y_train[test_idx]

    # Define pipeline
    pipeline = Pipeline([
        ('scaler', StandardScaler()),
        ('gpr', GaussianProcessRegressor(normalize_y=True))
    ])

    # GridSearchCV on pipeline
    inner_cv = GridSearchCV(
        estimator=pipeline,
        param_grid=param_grid,
        scoring='neg_mean_squared_error',
        cv=5,
        n_jobs=-1
    )
    inner_cv.fit(X_tr, y_tr)

    best_model = inner_cv.best_estimator_
    y_pred = best_model.predict(X_te)
    y_true_all.append(y_te[0])
    y_pred_all.append(y_pred[0])

# Final evaluation
y_true_all = np.array(y_true_all)
y_pred_all = np.array(y_pred_all)

r2_nested = r2_score(y_true_all, y_pred_all)
mse_nested = mean_squared_error(y_true_all, y_pred_all)
print(f"Nested LOOCV R²: {r2_nested:.4f}")
print(f"Nested LOOCV MSE: {mse_nested:.4f}")

# Plot
plt.figure(figsize=(5, 5))
plt.scatter(y_true_all, y_pred_all, alpha=0.7)
plt.plot([min(y_true_all), max(y_true_all)], [min(y_true_all), max(y_true_all)], 'r--', label=f'R2 = {r2_nested:.2f}')
plt.xlabel("True Efficiency")
plt.ylabel("Predicted Efficiency")
plt.title("Parity Plot - GPR (Nested LOOCV)")
plt.grid(True)
plt.tight_layout()
plt.legend()
plt.show()
```

There was a surprising result where for a small subset of the data, ridge regression and GPR with automatic relevance determination gave near perfect Parity plots when done on a subset of the data as above. Normally, I'd assume this was a bug, but the RFR wasn't a perfect parity. Please investigate. If you create a notebook, please include an Open in Colab link pointing to the specific notebook file from this branch (e.g., https://colab.research.google.com/github/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/official/model_monitoring/model_monitoring.ipynb) to make it easy for me to review.
